### PR TITLE
Refactored assembly to use in-memory model instead of IO model

### PR DESCRIPTION
**Pull request by fnqista**
The advantage of this code is streaming IO, with minimal reliance on disk speed. References to files and jar entries are stored as () => InputStream. Unzip to disk is gone, which is slow on some systems/hard drives.

Everything retained:
- all functionalities (shade, merge, caching) 
- reproducible build (entires built in a specific order, jar entry timestamps identical for all entries, hash consistent)

The following were new:
- No unzip
- I cannot rename directories, that is the advantage of the unzip function of the current implementation (since files are unzipped, parent renames will reflect automatically to children - I cannot do this efficiently with this implementation). For this implementation I can only rename files. If there is a project file that conflicts, the project name is appended (i.e. LICENSE_foo), where in the current implementation the folder named LICENSE is renamed instead. For dependencies, the old behavior was maintained (i.e.LICENSE_scala-library-2.13.3). Tests were updated to reflect this change
- Corrected the reporting of file merge counts
- I additionally check for conflicts of files with directories that have the same name and signal an error, which the current implementation does not (it only checks for LICENSE conflicts. The current implementation will only throw this error when the jar is being created. New tests were added to reflect this change
- The option to create the jar in parallel which is faster! (1.5 seconds vs 5 seconds) The catch is the reproducible build and caching won't work if the jar is built in parallel. The new property is AssemblyOption.reproducibleBuild
- Dropped cacheUnzip property in AssemblyOptions
- Dropped excludedFiles in AssemblyOptions (?) this is not documented and seems unused
- Caching is now implemented by checking the last update timestamps of all included dependencies, plus some of the properties from AssemblyOptions, and if the output jar exists already
- Documentation updates
- Merge strategy retains the same structure, but the implementation (inputs and outputs) have changed, to support the classes added (Dependency, Project, Library, MergedEntry, JarEntry)
- Newline appending Input Stream (called AppendEofInputStream) added in AssemblyUtils to support MergeStrategy.concat with newlines

### CODE REVIEW

File: .scalafmt.conf [hunk 1]
```diff
@@ -0,0 +1,34 @@
+version = "3.5.1"
```
Comment (eed3si9n) [inline .scalafmt.conf]: For the ease of reviewing, could we split Scalafmt into another PR?

File: README.md [hunk 2]
```diff
@@ -16,6 +16,14 @@ Reporting Issues & Contributing
...
 Before you email me, please read [Issue Reporting Guideline](CONTRIBUTING.md) carefully. Twice. (Don't email me)
 
+Major changes since 2.0.0
+-------------------------------
+The plugin has been refactored to use in-memory processing of library entries, in contrast to its old version where library jars are unzipped to disk.
+
+This has positive performance implications, especially for large projects, machines with slow disks (i.e. spinning hard drives) or systems with slow file systems (i.e. WSL1 emulated file access).
+
+It has also streamlined the plugin code for contributors.
+
```
Comment (eed3si9n) [inline README.md]: I suggest we move this part out to a new Google doc or gist, and eventually include it into GitHub Release note.
For reference the release note for sbt-assembly 1.0.0 looked like this - https://github.com/sbt/sbt-assembly/releases/tag/v1.0.0

Along with the new features, we should list out all breakages and document migration path where necessary since sbt-assembly is used by thousands of builds.

File: src/main/scala/sbtassembly/Assembly.scala [hunk 3]
```diff
@@ -1,394 +1,703 @@
...
+    if (mergedEntries.nonEmpty) {
+      mergedEntries
+        .groupBy(entry => entry.mergeStrategy.getClass.getName + "_" + entry.mergeStrategy.name)
+        .seq // we need to switch to sequential here to not mess up the detail logs
+        .foreach { case (strategyId, entries) =>
+          val mergeStrategy = entries.head.mergeStrategy
+          val entriesToNotify = entries.filter(entry => entry.origins.size >= mergeStrategy.notifyThreshold)
+          val totalMerged = entriesToNotify.map(_.origins.size).sum
+          if (totalMerged > 0) {
+            val custom = if (strategyId.startsWith(this.getClass.getPackageName)) "" else " (Custom)"
```
Comment (eed3si9n) [inline src/main/scala/sbtassembly/Assembly.scala]: Apparently this doesn't work on JDK 8

```
[error] /home/runner/work/sbt-assembly/sbt-assembly/src/main/scala/sbtassembly/Assembly.scala:605:66: value getPackageName is not a member of Class[T]
[error]             val custom = if (strategyId.startsWith(this.getClass.getPackageName)) "" else " (Custom)"
```

**Comment by eed3si9n** [comment]
@fnqista Thanks for this contribution. Looking forward to the perf improvements.
Since many Spark users use sbt-assembly and I am guessing that they use a lot of JARs, I wonder if we should do some before and after comparison to show case the difference in release notes etc.

File: src/main/scala/sbtassembly/Assembly.scala [hunk 4]
```diff
@@ -1,394 +1,703 @@
...
-          dest.mkdir()
-          AssemblyUtils.unzip(jar.data, dest, log)
-          IO.delete(ao.excludedFiles(Seq(dest)))
-          if (jarRules.nonEmpty) {
-            val mappings = ((dest ** (-DirectoryFilter)).get pair relativeTo(dest)) map {
-              case (k, v) => k.toPath -> v
-            }
-            Shader.shadeDirectory(dirRules, dest.toPath, mappings, ao.level == Level.Debug)
+    val clazzMappings =
+      timed(Level.Info, "Collect project classes") {
```
Comment (eed3si9n) [inline src/main/scala/sbtassembly/Assembly.scala]: Looking at the CI log, it shows:

```
[info] [info] Collect project classes start:
[info] [info] Collect project classes end. Took 0 ms
[info] [info] Collect dependency entries start:
[info] [info] Collect dependency entries end. Took 84 ms
[info] [info] Collect merge strategies start:
[info] [info] Collect merge strategies end. Took 29 ms
[info] [info] Merge all conflicting jar entries start:
[info] [info] 1 file(s) merged using strategy 'Discard' (Run the task at debug level to see the details)
[info] [info] Merge all conflicting jar entries end. Took 2378 ms
[info] [info] Finding remaining conflicts that were not merged start:
[info] [info] Finding remaining conflicts that were not merged end. Took 42 ms
[info] [info] Create jar start:
[info] [info] Create jar end. Took 606 ms
[info] [info] Jar hash: a6f5029aa2f99bc14c0a0461a0363d513c509bed
[info] [success] Total time: 3 s, completed Apr 16, 2022, 5:04:14 PM
```

If the entire operation is going to be really fast, and there's nothing actionable from the plugin users' point of view, I think we should remove all info level logs.

File: .scalafmt.conf (hunk 1 above)
Comment (fnqista) [reply .scalafmt.conf]: I will remove the scalafmt. I just needed it for formatting some of the files, but not all esp. AssemblyOption and AssemblyKeys.

@eed3si9n I kind of forgot, but when updating PRs, I should just do a new commit, right? Is it you who will take care of the squashing during merge?

File: README.md (hunk 2 above)
Comment (fnqista) [reply README.md]: I will remove it but I don't know how to do a new Google doc or gist yet. Is the code committed as part of this repo?

File: src/main/scala/sbtassembly/Assembly.scala (hunk 3 above)
Comment (fnqista) [reply src/main/scala/sbtassembly/Assembly.scala]: Fixing....

File: src/main/scala/sbtassembly/Assembly.scala (hunk 4 above)
Comment (fnqista) [reply src/main/scala/sbtassembly/Assembly.scala]: I will move all these timers levels to DEBUG

File: .scalafmt.conf (hunk 1 above)
Comment (eed3si9n) [reply .scalafmt.conf]: I am not too fussy about git history on plugin projects, so I'd probably keep the git history as-is if it makes sense to do so.

Comment (eed3si9n) [reply .scalafmt.conf]: if you want to stack up fixes on top, I can squash them on merge too.

File: README.md (hunk 2 above)
Comment (eed3si9n) [reply README.md]: > Is the code committed as part of this repo?

Not sure what you mean here.

Comment (fnqista) [reply README.md]: I mean the release notes :)

**Comment by fnqista** [comment]
> 

In my home computer with an NVME GEN4 drive (7000 MB/sec read) and an AMD Ryzen 5900X the difference between the old and the new implementation is not too much. I will run it on my work computer though, where we can see much more dramatic improvements.

File: README.md (hunk 2 above)
Comment (eed3si9n) [reply README.md]: Yes and no. Release notes https://github.com/sbt/sbt-assembly/releases is part of this _Github repo_, but it's not part of _git repo_ and there's no history tracking etc.

**Comment by fnqista** [comment]
> @fnqista Thanks for this contribution. Looking forward to the perf improvements. Since many Spark users use sbt-assembly and I am guessing that they use a lot of JARs, I wonder if we should do some before and after comparison to show case the difference in release notes etc.

When the computer is fast, like in my home computer which was upgraded recently, the difference between the old and new implementation is not that pronounced:

System: AMD Ryzen 5900X 12c/24t with NVME Gen4 drive (7000 MB/sec read)
Command: sbt clean assembly
Project: assembly jar ~90MB in size
- Old implementation 
`[success] Total time: 22 s, completed Apr 16, 2022, 10:04:32 PM`

- New implementation
`[success] Total time: 20 s, completed Apr 16, 2022, 10:13:26 PM`

However in my work computer, I have pre-tested before and something that takes 15 minutes in the old implementation only takes 29 seconds now. I will gather evidence when I am back at work after the Easter.

File: README.md [hunk 5]
```diff
@@ -436,6 +470,14 @@ lazy val app = (project in file("app"))
   )
 ```
 
+#### Jar assembly performance
+
```
Comment (eed3si9n) [inline README.md:496]: For ease of copy-pasting the following should be inserted here:

    ```scala
    ThisBuild / assemblyRepeatableBuild := false
    ```

**Comment by eed3si9n** [comment]
Here's a result from assembling a fresh Play app on my old mac laptop

| sbt-assembly  | assembly (median) | assembly (max out of 5) |
| ------------- | ------------- | ------------- |
| Baseline (1.2.0)               | 10s  | 12s |
| This PR (default)  | 7s  |  9s |
| This PR (`ThisBuild / assemblyRepeatableBuild := false`)  | 5s | 5s |

**Comment by fnqista** [comment]
@eed3si9n Release notes draft:
https://gist.github.com/fnqista/56fb6c5957302ba4af32c6c2b7dcb7c8

File: src/main/scala/sbtassembly/AssemblyPlugin.scala [hunk 6]
```diff
@@ -42,90 +44,80 @@ object AssemblyPlugin extends sbt.AutoPlugin {
...
-    assemblyPackageScala                            := Assembly.assemblyTask(assemblyPackageScala).value,
-    assembledMappings in assemblyPackageScala       := Assembly.assembledMappingsTask(assemblyPackageScala).value,
-    assemblyPackageDependency                       := Assembly.assemblyTask(assemblyPackageDependency).value,
-    assembledMappings in assemblyPackageDependency  := Assembly.assembledMappingsTask(assemblyPackageDependency).value,
+    assemblyPackageScala := Assembly.assemblyTask(assemblyPackageScala).value,
+    assemblyPackageDependency := Assembly.assemblyTask(assemblyPackageDependency).value,
 
     // test
-    test in assembly := { () },
+    test in assembly := Unit,
```
Comment (eed3si9n) [inline src/main/scala/sbtassembly/AssemblyPlugin.scala]: Is this `scala.Unit`, if so isn't that a companion object for `scala.Unit` type, and not an actual term of `Unit` type?

**Review by eed3si9n — CHANGES_REQUESTED** [review]

File: src/main/scala/sbtassembly/AssemblyPlugin.scala (hunk 6 above)
Comment (fnqista) [reply src/main/scala/sbtassembly/AssemblyPlugin.scala]: You are right. I changed this, silly, because scalafmt removes the surrounding braces, causing a compile error. I will find out how scalafmt can skip this. But for now, I will return it to what it was before.

Comment (fnqista) [reply src/main/scala/sbtassembly/AssemblyPlugin.scala]: Apparently we can use `test in assembly := {}`, and scalafmt leaves it alone. Is this fine?

Comment (eed3si9n) [reply src/main/scala/sbtassembly/AssemblyPlugin.scala]: I guess so.

**Comment by fnqista** [comment]
@eed3si9n The method signature of `MergeStrategy.merge` also has changed:

`merge(conflicts: Vector[Dependency]): Either[String, Vector[JarEntry]]` whereas the old one just uses standard primitive and collection types.

This is exposed to the user because they can create custom merge strategies. Shall we use contraband for these to at least guarantee some level of backward compatibility?

File: src/main/scala/sbtassembly/MergeStrategy.scala [hunk 7]
```diff
@@ -1,148 +1,245 @@
...
 
 /**
- * MergeStrategy is invoked if more than one source file is mapped to the 
- * same target path. Its arguments are the tempDir (which is deleted after
- * packaging) and the sequence of source files, and it shall return the
- * file to be included in the assembly (or throw an exception).
+ * Class for merging files that shares the same target path
  */
-abstract class MergeStrategy extends Function1[(File, String, Seq[File]), Either[String, Seq[(File, String)]]] {
+trait MergeStrategy {
```
Comment (eed3si9n) [inline src/main/scala/sbtassembly/MergeStrategy.scala]: > @eed3si9n The method signature of `MergeStrategy.merge` also has changed:
> 
> `merge(conflicts: Vector[Dependency]): Either[String, Vector[JarEntry]]` whereas the old one just uses standard primitive and collection types.
> 
> This is exposed to the user because they can create custom merge strategies. Shall we use contraband for these to at least guarantee some level of backward compatibility?

Thanks for highlighting this point. I think extending `Function1` is cooler than making `def merge`. If this is mostly implementation detail, I feel like we can do some fp here.

If we provide some kind of function transformer as `MergeStrategy.apply(...)` that takes old function, and returns a new `MergeStrategy` that would be great.

Comment (fnqista) [reply src/main/scala/sbtassembly/MergeStrategy.scala]: > Thanks for highlighting this point. I think extending `Function1` is cooler than making `def merge`. If this is mostly implementation detail, I feel like we can do some fp here.

Well, in my experience, when I create a custom MergeStrategy in 1.2.0, it is _not very clear_ to me what I need to do as it is a function.
The method signature 

```scala
Function1[(File, String, Seq[File]), Either[String, Seq[(File, String)]]]
```
is confusing since a function does not name its parameters, unless I look at one of the existing implementations, where I can learn that the first argument is the `tempDir` - but it still is not clear how to use it or if I even need it.
The other parameters maybe a little bit clearer once one looks at the implementation.

```scala
val first: MergeStrategy = new MergeStrategy {
    val name = "first"
    def apply(tempDir: File, path: String, files: Seq[File]): Either[String, Seq[(File, String)]] =
      Right(Seq(files.head -> path))
  }
```

With the changed signature, it is very clear that I am given a trait to implement and I have to override the `merge` and the `name` methods.

```scala
  case _ =>
    new MergeStrategy {
      override def name = "First"

      override def merge(conflicts: Vector[Assembly.Dependency]) =
        Right(Vector(JarEntry(conflicts.head.target, conflicts.head.stream)))
    }
```
That is my main motivation for changing this part of the code - to make it easier to understand for users.

> If we provide some kind of function transformer as `MergeStrategy.apply(...)` that takes old function, and returns a new `MergeStrategy` that would be great.

I cannot retrofit the **old** function to the new one, as `File` represents **extracted files** when we talk of jars. For project classes, this is easy to make but then we lose uniformity of the signature with libraries.

The new `Dependency` type instead carries a lazy input stream payload of `() => InputStream` that can refer to the **content of an entry inside an _unextracted_ jar** plus of course the `target` path. If people wanted to find out `sourceOfFiileForMerge` in the old implementation, they can extract the `module: Option[ModuleCoordinate]` for jars or the `source` for project classes.

So, if possible, I wanted to keep the change, and maybe to increase backward compatibility for 2.x.x, I could learn to use `contraband` instead of case classes.

But, IMHO, the old MergeStrategy contract/semantics is very hard to understand.

Comment (eed3si9n) [reply src/main/scala/sbtassembly/MergeStrategy.scala]: I see that the old impl may have been confusing, but with `Vector[Dependency]` being the only input now for `merge` isn't the semantics of `Vector[Dependency] => Either[String, Vector[JarEntry]]` clearer now?
Instead of `new MergeStrategy` we can create `apply(name: String)(f: Vector[Dependency] => Either[String, Vector[JarEntry]])` on the companion object so a custom strategy could look like:

```scala
MergeStrategy(name = "First") { conflicts =>
  Right(Vector(JarEntry(conflicts.head.target, conflicts.head.stream)))
}
```

re: Contraband, I don't see a huge upside to using it here since Contraband is basically a case class replacement that's capable of keeping binary compatibility and persisting into JSON.

Comment (fnqista) [reply src/main/scala/sbtassembly/MergeStrategy.scala]: > I see that the old impl may have been confusing, but with `Vector[Dependency]` being the only input now for `merge` isn't the semantics of `Vector[Dependency] => Either[String, Vector[JarEntry]]` clearer now? Instead of `new MergeStrategy` we can create `apply(name: String)(f: Vector[Dependency] => Either[String, Vector[JarEntry]])` on the companion object so a custom strategy could look like:
> 
> ```scala
> MergeStrategy(name = "First") { conflicts =>
>   Right(Vector(JarEntry(conflicts.head.target, conflicts.head.stream)))
> }
> ```
> 
> re: Contraband, I don't see a huge upside to using it here since Contraband is basically a case class replacement that's capable of keeping binary compatibility and persisting into JSON.

Ahh I see what you mean now. This is definitely better. I just got confused when you mentioned "old" because even the old implementation used `new MergeStrategy(...)` for built-in merge strats.

It makes sense now

**Comment by fnqista** [comment]
@eed3si9n 
Major PR updates:
- As agreed, I changed the implementation of `MergeStrategy` as per your suggestion:

```scala
val first: MergeStrategy = MergeStrategy("First") { conflicts =>
  Right(Vector(JarEntry(conflicts.head.target, conflicts.head.stream)))
}
```
- For custom merge strategies (user-provided):
```scala
ThisBuild / assemblyMergeStrategy := {
...
case _ => CustomMergeStrategy("First") { conflicts =>
  Right(Vector(JarEntry(conflicts.head.target, conflicts.head.stream)))
}

```
- I have to change the object for a custom merge strategy because:
  - Primarily to namespace them. It is easy for a user to create a custom merge strategy that is also named `First`, but properly namespacing them makes us distinguish each other when needed in both the code and in the logs.
  - To prevent future plugin contributors who will create/modify a built-in MergeStrategy in as a "custom" one, again breaking namespacing.
 You can browse the code under `sbtassembly.MergeStrategy`. I used a sealed trait that extends a `Function1` plus two descendants to represent the namespacing. Companion objects are also created for them.
 
 - And another one - if the user includes `.class` files in a built-in `rename` strategy, I throw an error now (well, really, return a `Left`), and inform them that they should use `shadeRules` instead. In 1.2.0 this was ignored but also silently a no-op, without informing the user. I think it is important to be transparent with the plugin behavior regarding this. Code is in `MergeStrategy`, plus a new test is added for this.

- I updated the `README` to clarify that a `MergeStrategy` doesn't only merge, it also transforms, such as in the case of a rename, where it probably matches only a single file. I think it is important that this is communicated to the users.

- I exposed `JarEntry` and `CustomMergeStrategy` as an auto import in `AssemblyPlugin`, since I have tried this myself and found out that this increases the user's convenience.

- I also created an implicit class conversion from `String -> java.nio.file.Path` since `Path` is used by `JarEntry`, again in `AssemblyPlugin`.

- I implemented the other review comment about `test in assembly := {},`, which now properly returns `Unit`, again in `AssemblyPlugin`.

- I type-aliased `() => InputStream` as `LazyInputStream` in `Assembly`

- I finally moved the `rename` logic inside the `cache` block too. Now, to check the cache, nothing is every really done except to build the `Dependency` to check if their `timestamp`s have changed and to build a `Map` of paths to merge strategies. Before, we always process the renames before building the merge strategy map/checking the cache. I don't know what I was thinking before, but it seems that we don't always need to process `rename`s before checking the cache (I hope I am right). Anyway, I tested it quite a few times and it seems solid. Code is in `Assembly`.

- I created the `MergeStrategy.isBuiltIn` property to namespace custom ones from, well, built-in ones - and make sure it is included in the cache checking. Also in `Assembly`.

- Minor code and test updates due to the above changes.

File: src/main/scala/sbtassembly/Assembly.scala [hunk 8]
```diff
@@ -1,394 +1,715 @@
...
-    import Tracked.{ inputChanged, outputChanged }
-    import Cache._
-    import FileInfo.{hash, exists}
-    import java.util.jar.{Attributes, Manifest}
+  /**
+   * Represents an assembly jar entry
+   * @param target the path to write in the jar
+   * @param stream the byte payload
+   */
+  case class JarEntry(target: Path, stream: LazyInputStream) {
```
Comment (eed3si9n) [inline src/main/scala/sbtassembly/Assembly.scala]: Traditionally we use `(File, String)` pair or `Seq` of that as "mapping" - https://www.scala-sbt.org/1.x/docs/Mapping-Files.html

Using NIO `Path` might throw people off since it _sounds_ more specific than `String`, but `Path` would now be local operating system's path implementation (including Windows), and zip entry to `C:\foo\` won't make sense. So it's a mix bag because it's better typed, and worse typed at the same time.

Comment (fnqista) [reply src/main/scala/sbtassembly/Assembly.scala]: Hey Eugene, thanks for the comments.

Actually, I am aware of this now while building the plug-in, but `Path` actually makes sense. If you look at the `createJar` code, you can actually mount a jar/zip file as an NIO file system:

```scala
jarFileSystemResource(URI.create(s"jar:file:${output.toPath.toString}")) { jarFs =>
      val manifestPath = jarFs.getPath("META-INF/MANIFEST.MF")
```

`jarFs` is actually an NIO `FileSystem` and `jarFs.getPath` nets an NIO `Path`.

(It's a neat trick I learned and borrowed from `mill`'s assembly code.)

So  - we _can_ represent zip/jar entries as a path.

EDIT:
Code for the `jarFileSystemResource`:

```scala
private[sbtassembly] val jarFileSystemResource =
    Using.resource((uri: URI) => FileSystems.newFileSystem(uri, Map("create" -> "true").asJava))
```

Comment (eed3si9n) [reply src/main/scala/sbtassembly/Assembly.scala]: Yea. I am aware of zipfs and NIO's ability to represent JAR entries (we borrow some of the code in Zinc to update JAR incrementally), but `Path` as a type is superset of various paths, so if a function (or in this case a case class) accepts `Path`, wouldn't it sort of imply it can handle _all_ paths? Also a `Path` representation of JAR-path would include the name of the zip or JAR file in the file system?, which too could be inconsistent with what we need, which is (relative) path entry. In other words if someone specifies `JarEntry(convertToPath(URI.create(s"jar:file:/tmp/some.jar!/foo/LICENSE")), conflicts.head.stream)`, and we're not dealing with `/tmp/some.jar`, what should sbt-assembly do? Why should we make that state representable?

Comment (fnqista) [reply src/main/scala/sbtassembly/Assembly.scala]: @eed3si9n 
I understand where you come from. I integrated with AWS S3's API and representing a Path is a pain. String is too generic and type-unsafe, everything can be a String. NIO Path is, as you've said above...

I am not aware of any path library that is not tied to any OS as well (such as Amazon's S3, whose root path is an empty string instead of `/`. I actually had to hand-roll my own Path implementation in that project handled via type classes to distinguish root paths of differnt file systems - but that's another story.

So this is a dilemma. 

So - what do you think we should do? Use String anyway?

Comment (eed3si9n) [reply src/main/scala/sbtassembly/Assembly.scala]: I guess my vote would be to keep it as String.

Related, sbt-assembly includes `PathList` that helps users pattern match String as delimiter-delimited Strings - https://github.com/sbt/sbt-assembly/blob/c86a0f7c1c4edd435e8c512f2716a71ed9ce8b7f/src/main/scala/sbtassembly/Assembly.scala#L387-L394

somewhat documented as https://github.com/sbt/sbt-assembly#merge-strategy

**Review by eed3si9n — APPROVED** [review]
LGTM

**Comment by fnqista** [comment]
@eed3si9n 
So, me and a few colleagues tested this at work over the past week and we found some bugs, all of them related to custom merge strategies.

- The Custom rename strategy method signature must be changed so it can only return a `String` representing the renamed target, instead of `Vector[JarEntry]`, to prevent users from corrupting the result, when they should only be renaming the `target`. So the correct signature should be `CustomStrategy.rename(f: Dependency => String)` instead of the old `CustomStrategy.rename(f: Vector[Dependency] => Vector[Either[String, JarEntry]]`)

- Renames should really be done before evaluating the cache, which I broke in my previous commit. Otherwise, if we do it after checking the cache, the assembly might not execute when it should, in cases where the rename might cause another conflict.

- Custom merge strategies, other than rename, breaks the caching. In this PR, the only way to check if the merge strategies have changed is by creating a `Map[String, String]` representing the matched `target` and the merge strategy `name`. This is mostly ok for built-in strategies and the custom rename because their code/logic doesn't change. But with custom merge strategies, the `name` means nothing when caching since we cannot predict what the user will do, thereby it breaks caching again by the code thinking that nothing has changed when in fact the user updated the merge strategy logic and assembly should re-run.

I have prepared some fixes but I am still testing them, hopefully I can make a new push within a couple of days.

EDIT: Clarifying point no. 3 about custom renames being OK - If we fix no. 2 (rename before caching), then custom renames will happen before the cache.

**Comment by fnqista** [comment]
> @eed3si9n So, me and a few colleagues tested this at work over the past week and we found some bugs, all of them related to custom merge strategies.
> 
> * The Custom rename strategy method signature must be changed so it can only return a `String` representing the renamed target, instead of `Vector[JarEntry]`, to prevent users from corrupting the result, when they should only be renaming the `target`. So the correct signature should be `CustomStrategy.rename(f: Dependency => String)` instead of the old `CustomStrategy.rename(f: Vector[Dependency] => Vector[Either[String, JarEntry]]`)
> * Renames should really be done before evaluating the cache, which I broke in my previous commit. Otherwise, if we do it after checking the cache, the assembly might not execute when it should, in cases where the rename might cause another conflict.
> * Custom merge strategies, other than rename, breaks the caching. In this PR, the only way to check if the merge strategies have changed is by creating a `Map[String, String]` representing the matched `target` and the merge strategy `name`. This is mostly ok for built-in strategies and the custom rename because their code/logic doesn't change. But with custom merge strategies, the `name` means nothing when caching since we cannot predict what the user will do, thereby it breaks caching again by the code thinking that nothing has changed when in fact the user updated the merge strategy logic and assembly should re-run.
> 
> I have prepared some fixes but I am still testing them, hopefully I can make a new push within a couple of days.
> 
> EDIT: Clarifying point no. 3 about custom renames being OK - If we fix no. 2 (rename before caching), then custom renames will happen before the cache.

Fixes committed for the above...hopefully the last one.

EDIT: Also updated draft release notes: https://gist.github.com/fnqista/56fb6c5957302ba4af32c6c2b7dcb7c8

**Comment by fnqista** [comment]
@eed3si9n Can you review the pending changes and then create an RC1 candidate when you have time?

**Review by eed3si9n — APPROVED** [review]
If the test passes I am good here.
